### PR TITLE
Ensure all gulp tasks run to completion

### PR DESCRIPTION
**Pull request by colinrotherham**
This PR ensures all Gulp tasks returns either a Promise or `gulp.src()` stream

The CLI wasn't waiting for tasks to run to completion, ~including a “write after end” YAML error~ see https://github.com/alphagov/govuk-frontend/pull/2870

**Fixes this workaround**
```js
// Required for Gulp task to complete successfully.
// Can be removed once this no longer runs via Gulp.
cb()
```

### CODE REVIEW

**Comment by 36degrees** [comment]
For some reason, I'm intermittently missing the warning text fixtures.json when running `npm run build:package` (flagged by tests, but verified the file is actually missing after the task has completed):

```
➜ npm run build:package

> build:package
> gulp build:package --destination 'package' && npm run test:build:package

[11:24:56] Using gulpfile ~/Code/govuk-frontend/gulpfile.js
[11:24:56] Starting 'build:package'...
[11:24:56] Starting 'cleanPackage'...
[11:24:56] Finished 'cleanPackage' after 52 ms
[11:24:56] Starting 'copy-files'...
[11:24:56] Finished 'copy-files' after 526 ms
[11:24:56] Starting 'js:compile'...
[11:24:56] Finished 'js:compile' after 28 ms
[11:24:56] Starting 'js:copy-esm'...
[11:24:57] Finished 'js:copy-esm' after 1.07 s
[11:24:57] Finished 'build:package' after 1.68 s

> test:build:package
> npx jest --selectProjects "Gulp tasks" --testMatch "**/*build-package*"

Running one project: Gulp tasks
 FAIL   Gulp tasks  tasks/gulp/__tests__/after-build-package.test.js
  package/
    ✓ should contain the expected files (1 ms)
    README.md
      ✓ is not overwritten (6 ms)
    all.scss
      ✓ should compile without throwing an exception (130 ms)
    all.js
      ✓ should have correct module name (1 ms)
    component
      ✓ 'accordion' should have macro-options.json that contains JSON (1 ms)
      ✓ 'back-link' should have macro-options.json that contains JSON (2 ms)
      ✓ 'breadcrumbs' should have macro-options.json that contains JSON
      ✓ 'button' should have macro-options.json that contains JSON (1 ms)
      ✓ 'character-count' should have macro-options.json that contains JSON (1 ms)
      ✓ 'checkboxes' should have macro-options.json that contains JSON
      ✓ 'cookie-banner' should have macro-options.json that contains JSON
      ✓ 'date-input' should have macro-options.json that contains JSON (1 ms)
      ✓ 'details' should have macro-options.json that contains JSON (1 ms)
      ✓ 'error-message' should have macro-options.json that contains JSON
      ✓ 'error-summary' should have macro-options.json that contains JSON (1 ms)
      ✓ 'fieldset' should have macro-options.json that contains JSON
      ✓ 'file-upload' should have macro-options.json that contains JSON
      ✓ 'footer' should have macro-options.json that contains JSON (1 ms)
      ✓ 'header' should have macro-options.json that contains JSON (1 ms)
      ✓ 'hint' should have macro-options.json that contains JSON (1 ms)
      ✓ 'input' should have macro-options.json that contains JSON
      ✓ 'inset-text' should have macro-options.json that contains JSON
      ✓ 'label' should have macro-options.json that contains JSON (1 ms)
      ✓ 'notification-banner' should have macro-options.json that contains JSON
      ✓ 'pagination' should have macro-options.json that contains JSON (1 ms)
      ✓ 'panel' should have macro-options.json that contains JSON
      ✓ 'phase-banner' should have macro-options.json that contains JSON (1 ms)
      ✓ 'radios' should have macro-options.json that contains JSON (1 ms)
      ✓ 'select' should have macro-options.json that contains JSON
      ✓ 'skip-link' should have macro-options.json that contains JSON (1 ms)
      ✓ 'summary-list' should have macro-options.json that contains JSON
      ✓ 'table' should have macro-options.json that contains JSON (1 ms)
      ✓ 'tabs' should have macro-options.json that contains JSON
      ✓ 'tag' should have macro-options.json that contains JSON (1 ms)
      ✓ 'textarea' should have macro-options.json that contains JSON (1 ms)
      ✓ 'warning-text' should have macro-options.json that contains JSON
    components with JavaScript
      ✓ 'package/govuk/components/accordion/accordion.js' should have component JavaScript file with correct module name (1 ms)
      ✓ 'package/govuk/components/button/button.js' should have component JavaScript file with correct module name
      ✓ 'package/govuk/components/character-count/character-count.js' should have component JavaScript file with correct module name (1 ms)
      ✓ 'package/govuk/components/checkboxes/checkboxes.js' should have component JavaScript file with correct module name
      ✓ 'package/govuk/components/details/details.js' should have component JavaScript file with correct module name
      ✓ 'package/govuk/components/error-summary/error-summary.js' should have component JavaScript file with correct module name (1 ms)
      ✓ 'package/govuk/components/header/header.js' should have component JavaScript file with correct module name
      ✓ 'package/govuk/components/notification-banner/notification-banner.js' should have component JavaScript file with correct module name
      ✓ 'package/govuk/components/radios/radios.js' should have component JavaScript file with correct module name
      ✓ 'package/govuk/components/skip-link/skip-link.js' should have component JavaScript file with correct module name
      ✓ 'package/govuk/components/tabs/tabs.js' should have component JavaScript file with correct module name (1 ms)
    fixtures
      ✓ 'accordion' should have fixtures.json that contains JSON (1 ms)
      ✓ 'back-link' should have fixtures.json that contains JSON
      ✓ 'breadcrumbs' should have fixtures.json that contains JSON (1 ms)
      ✓ 'button' should have fixtures.json that contains JSON (2 ms)
      ✓ 'character-count' should have fixtures.json that contains JSON (1 ms)
      ✓ 'checkboxes' should have fixtures.json that contains JSON (2 ms)
      ✓ 'cookie-banner' should have fixtures.json that contains JSON (2 ms)
      ✓ 'date-input' should have fixtures.json that contains JSON (1 ms)
      ✓ 'details' should have fixtures.json that contains JSON (6 ms)
      ✓ 'error-message' should have fixtures.json that contains JSON (1 ms)
      ✓ 'error-summary' should have fixtures.json that contains JSON (1 ms)
      ✓ 'fieldset' should have fixtures.json that contains JSON (1 ms)
      ✓ 'file-upload' should have fixtures.json that contains JSON (1 ms)
      ✓ 'footer' should have fixtures.json that contains JSON (1 ms)
      ✓ 'header' should have fixtures.json that contains JSON (2 ms)
      ✓ 'hint' should have fixtures.json that contains JSON (2 ms)
      ✓ 'input' should have fixtures.json that contains JSON (2 ms)
      ✓ 'inset-text' should have fixtures.json that contains JSON
      ✓ 'label' should have fixtures.json that contains JSON (1 ms)
      ✓ 'notification-banner' should have fixtures.json that contains JSON (1 ms)
      ✓ 'pagination' should have fixtures.json that contains JSON (1 ms)
      ✓ 'panel' should have fixtures.json that contains JSON
      ✓ 'phase-banner' should have fixtures.json that contains JSON (1 ms)
      ✓ 'radios' should have fixtures.json that contains JSON (3 ms)
      ✓ 'select' should have fixtures.json that contains JSON (1 ms)
      ✓ 'skip-link' should have fixtures.json that contains JSON (7 ms)
      ✓ 'summary-list' should have fixtures.json that contains JSON (2 ms)
      ✓ 'table' should have fixtures.json that contains JSON (1 ms)
      ✓ 'tabs' should have fixtures.json that contains JSON (1 ms)
      ✓ 'tag' should have fixtures.json that contains JSON (1 ms)
      ✓ 'textarea' should have fixtures.json that contains JSON (1 ms)
      ✕ 'warning-text' should have fixtures.json that contains JSON

  ● package/ › fixtures › 'warning-text' should have fixtures.json that contains JSON

    ENOENT: no such file or directory, open 'package/govuk/components/warning-text/fixtures.json'



Test Suites: 1 failed, 1 total
Tests:       1 failed, 78 passed, 79 total
Snapshots:   0 total
Time:        0.856 s, estimated 1 s
Ran all test suites.
```

(I have run `npm install` so should be using the latest version of `js-yaml` as intended)

**Comment by 36degrees** [comment]
Also seeing issues running `npm build:dist`:

```
➜ npm run build:dist              

> build:dist
> gulp build:dist --destination 'dist' && npm run test:build:dist

[11:29:05] Using gulpfile ~/Code/govuk-frontend/gulpfile.js
[11:29:05] Starting 'build:dist'...
[11:29:05] Starting 'cleanDist'...
[11:29:05] Finished 'cleanDist' after 22 ms
[11:29:05] Starting 'copy-assets'...
[11:29:05] Starting 'styles'...
[11:29:05] Starting 'scss:lint'...
[11:29:08] Finished 'scss:lint' after 2.26 s
[11:29:08] Starting 'scss:compile'...
[11:29:08] Finished 'scss:compile' after 361 ms
[11:29:08] Finished 'styles' after 2.62 s
[11:29:08] Starting 'scripts'...
[11:29:08] Starting 'js:lint'...

/Users/oliver.byford/Code/govuk-frontend/src/govuk/i18n.unit.test.mjs
  182:37  warning  Expected property shorthand  object-shorthand
  194:37  warning  Expected property shorthand  object-shorthand
  206:37  warning  Expected property shorthand  object-shorthand
  218:37  warning  Expected property shorthand  object-shorthand
  230:37  warning  Expected property shorthand  object-shorthand
  242:37  warning  Expected property shorthand  object-shorthand
  254:37  warning  Expected property shorthand  object-shorthand
  266:37  warning  Expected property shorthand  object-shorthand
  278:37  warning  Expected property shorthand  object-shorthand

✖ 9 problems (0 errors, 9 warnings)
  0 errors and 9 warnings potentially fixable with the `--fix` option.


[11:29:09] Finished 'js:lint' after 1.25 s
[11:29:09] Starting 'js:compile'...
[11:29:09] Finished 'js:compile' after 2.05 ms
[11:29:09] Finished 'scripts' after 1.25 s
[11:29:09] Finished 'copy-assets' after 3.88 s
[11:29:09] Starting 'copy:assets'...
[11:29:10] The following tasks did not complete: build:dist, copy:assets
[11:29:10] Did you forget to signal async completion?

node:events:491
      throw er; // Unhandled 'error' event
      ^
PluginError [GulpUglifyError]: unable to minify JavaScript
    at createError (/Users/oliver.byford/Code/govuk-frontend/node_modules/gulp-uglify/lib/create-error.js:5:14)
    at /Users/oliver.byford/Code/govuk-frontend/node_modules/gulp-uglify/lib/minify.js:56:15
    at DestroyableTransform._transform (/Users/oliver.byford/Code/govuk-frontend/node_modules/gulp-uglify/composer.js:12:19)
    at DestroyableTransform.Transform._read (/Users/oliver.byford/Code/govuk-frontend/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:184:10)
    at DestroyableTransform.Transform._write (/Users/oliver.byford/Code/govuk-frontend/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:172:83)
    at doWrite (/Users/oliver.byford/Code/govuk-frontend/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:428:64)
    at writeOrBuffer (/Users/oliver.byford/Code/govuk-frontend/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:417:5)
    at DestroyableTransform.Writable.write (/Users/oliver.byford/Code/govuk-frontend/node_modules/through2/node_modules/readable-stream/lib/_stream_writable.js:334:11)
    at GulpRollup.ondata (node:internal/streams/readable:754:22)
    at GulpRollup.emit (node:events:513:28)
Emitted 'error' event on Domain instance at:
    at GulpRollup.emit (node:domain:540:12)
    at GulpRollup.onerror (/Users/oliver.byford/Code/govuk-frontend/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:640:52)
    at GulpRollup.emit (node:events:513:28)
    at GulpRollup.emit (node:domain:489:12)
    at /Users/oliver.byford/Code/govuk-frontend/node_modules/gulp-better-rollup/index.js:157:11
    at processTicksAndRejections (node:internal/process/task_queues:78:11) {
  cause: JS_Parse_Error [SyntaxError]: Unexpected token: name «placeholderValue», expected: punc «;»
      at JS_Parse_Error.get (eval at <anonymous> (/Users/oliver.byford/Code/govuk-frontend/node_modules/uglify-js/tools/node.js:18:1), <anonymous>:71:23)
      at getStackString (node:internal/util/inspect:1221:37)
      at formatError (node:internal/util/inspect:1343:15)
      at formatRaw (node:internal/util/inspect:970:14)
      at formatValue (node:internal/util/inspect:825:10)
      at formatProperty (node:internal/util/inspect:1892:11)
      at formatRaw (node:internal/util/inspect:1038:9)
      at formatValue (node:internal/util/inspect:825:10)
      at inspect (node:internal/util/inspect:350:10)
      at afterInspector (node:internal/errors:794:14) {
    filename: 'all.mjs',
    line: 577,
    col: 10,
    pos: 19976
  },
  plugin: 'gulp-better-rollup',
  fileName: '/Users/oliver.byford/Code/govuk-frontend/src/govuk/all.mjs',
  showStack: false,
  domainEmitter: GulpRollup {
    _readableState: ReadableState {
      objectMode: true,
      highWaterMark: 16,
      buffer: BufferList { head: null, tail: null, length: 0 },
      length: 0,
      pipes: [],
      flowing: false,
      ended: false,
      endEmitted: false,
      reading: false,
      constructed: true,
      sync: false,
      needReadable: true,
      emittedReadable: false,
      readableListening: false,
      resumeScheduled: false,
      errorEmitted: false,
      emitClose: true,
      autoDestroy: true,
      destroyed: false,
      errored: null,
      closed: false,
      closeEmitted: false,
      defaultEncoding: 'utf8',
      awaitDrainWriters: null,
      multiAwaitDrain: false,
      readingMore: false,
      dataEmitted: true,
      decoder: null,
      encoding: null,
      [Symbol(kPaused)]: true
    },
    _events: [Object: null prototype] { prefinish: [Function: prefinish] },
    _eventsCount: 1,
    _maxListeners: undefined,
    _writableState: WritableState {
      objectMode: true,
      highWaterMark: 16,
      finalCalled: false,
      needDrain: false,
      ending: true,
      ended: true,
      finished: false,
      destroyed: false,
      decodeStrings: true,
      defaultEncoding: 'utf8',
      length: 1,
      writing: true,
      corked: 0,
      sync: false,
      bufferProcessing: false,
      onwrite: [Function: bound onwrite],
      writecb: [Function: nop],
      writelen: 1,
      afterWriteTickInfo: null,
      buffered: [],
      bufferedIndex: 0,
      allBuffers: true,
      allNoop: true,
      pendingcb: 1,
      constructed: true,
      prefinished: false,
      errorEmitted: false,
      emitClose: true,
      autoDestroy: true,
      errored: null,
      closed: false,
      closeEmitted: false,
      [Symbol(kOnFinished)]: []
    },
    allowHalfOpen: true,
    arg1: {
      name: 'GOVUKFrontend',
      legacy: true,
      format: 'umd',
      file: undefined,
      exports: undefined,
      globals: undefined,
      amd: { id: 'GOVUKFrontend' },
      sourcemap: false
    },
    arg2: undefined,
    [Symbol(kCapture)]: false,
    [Symbol(kCallback)]: null
  },
  domainThrown: false,
  showProperties: true
}
```

**Comment by colinrotherham** [comment]
@36degrees I've resolved those two issues:

1. **Missing `fixtures.json` output**
The `gulp copy-files` task relied on the `generateFixtures()` mapper running "out-of-order" (callback fired early) which annoyingly [hits an obscure `gulp-rename` bug](https://github.com/gulpjs/gulp/issues/1508) when resolved https://github.com/alphagov/govuk-frontend/blob/bdd019b1171f8b26faa3a356bd00fd772b049135/tasks/gulp/copy-to-destination.js#L35

2. **Unable to minify JavaScript for `./dist`**
The `GulpUglifyError` affects `main` but I've fixed it in https://github.com/alphagov/govuk-frontend/pull/2868

File: tasks/gulp/lint.js [hunk 1]
```diff
@@ -1,21 +1,41 @@
...
-gulp.task('scss:lint', (done) => {
+/**
+ * Spawns Node.js child process for npm task
+ * rather than using Gulp
+ *
+ * @param {string} name - npm script name
+ * @param {string[]} [args] - npm script arguments
+ * @returns {Promise<number>} Exit code
+ */
+async function runTask (name, args = []) {
```
Comment (domoscargin) [inline tasks/gulp/lint.js:26]: ❤️

**Comment by colinrotherham** [comment]
Thanks @36degrees @domoscargin

Thinking it would help get this reviewed, I've moved some commits to a separate PR:
[**#2870** Fix stray let in I18n preventing uglify (again)](https://github.com/alphagov/govuk-frontend/pull/2870)

Would you mind please approving there as well?

Helps describe each piece of the puzzle a bit better

**Comment by colinrotherham** [comment]
This PR was related to https://github.com/alphagov/govuk-frontend/issues/2243

**Pull request merged**